New packages: highway, libjxl, qt-jpegxl-image-plugin

joshuakraemer opened the pull request:
#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me

These packages provide support for the JPEG XL image format.

closes #40552

mustaqimM commented:
Thanks for this PR, @joshuakraemer. I've updated it to the latest version and I'm busy using it. Hopefully it can get merged

```diff
diff --git a/common/shlibs b/common/shlibs
index c0d304de40..02392241b9 100644
--- a/common/shlibs
+++ b/common/shlibs
@@ -4038,5 +4038,5 @@ libstemmer.so.2 libstemmer-2.1.0_1
 libdecor-0.so.0 libdecor-0.1.0_1
 libavif.so.13 libavif-0.9.3_1
 libkdumpfile.so.9 libkdumpfile-0.4.1_1
-libjxl.so.0 libjxl-0.5.0_1
-libjxl_threads.so.0 libjxl-0.5.0_1
+libjxl.so.0.6 libjxl-0.6.1_1
+libjxl_threads.so.0.6 libjxl-0.6.1_1
diff --git a/srcpkgs/highway/template b/srcpkgs/highway/template
index 462df78a48..14cec1065d 100644
--- a/srcpkgs/highway/template
+++ b/srcpkgs/highway/template
@@ -1,6 +1,6 @@
 # Template file for 'highway'
 pkgname=highway
-version=0.12.2
+version=0.14.2
 revision=1
 build_style=cmake
 configure_args="-DBUILD_TESTING=OFF"
@@ -9,4 +9,4 @@ maintainer="Joshua Krämer <joshua@kraemer.link>"
 license="Apache-2.0"
 homepage="https://github.com/google/highway"
 distfiles="https://github.com/google/highway/archive/${version}.tar.gz"
-checksum=e1d47ce510429fdcf31f41697ca74fb0dcd59d933196e531a86d51751a56f4cc
+checksum=58f7f3a12394341796fbb3045d18020474cb214ef85b9a50bbaaa809e34e9232
diff --git a/srcpkgs/libjxl/template b/srcpkgs/libjxl/template
index 97df887973..14338d6d9b 100644
--- a/srcpkgs/libjxl/template
+++ b/srcpkgs/libjxl/template
@@ -1,9 +1,9 @@
 # Template file for 'libjxl'
 pkgname=libjxl
-version=0.5.0
+version=0.6.1
 revision=1
-_short_version=0.5
-wrksrc=libjxl-${_short_version}
+# _short_version=0.5
+wrksrc=libjxl-${version}
 build_style=cmake
 configure_args="-DBUILD_TESTING=OFF -DJPEGXL_ENABLE_BENCHMARK=OFF -DJPEGXL_ENABLE_EXAMPLES=OFF -DJPEGXL_ENABLE_SJPEG=OFF -DJPEGXL_ENABLE_PLUGINS=ON"
 hostmakedepends="tar pkg-config asciidoc"
@@ -12,13 +12,13 @@ short_desc="JPEG XL image format reference implementation"
 maintainer="Joshua Krämer <joshua@kraemer.link>"
 license="BSD-3-Clause, custom:Patent grant"
 homepage="https://github.com/libjxl/libjxl"
-_lodepng_hash=48e5364ef48ec2408f44c727657ac1b6703185f8
+_lodepng_hash=8c6a9e30576f07bf470ad6f09458a2dcd7a6a84a
 _skcms_hash=64374756e03700d649f897dbd98c95e78c30c7da
-distfiles="https://github.com/libjxl/libjxl/archive/v${_short_version}.tar.gz
+distfiles="https://github.com/libjxl/libjxl/archive/v${version}.tar.gz
  https://github.com/lvandeve/lodepng/archive/${_lodepng_hash}.tar.gz>lodepng-${_lodepng_hash}.tar.gz
  https://skia.googlesource.com/skcms/+archive/${_skcms_hash}.tar.gz>skcms-${_skcms_hash}.tar.gz"
-checksum="911cb4b50eb621131ca22382166f40d4914a4ff4453dd299ade1e3292f311f89
- c47c48c77a205f1af484b7b5a847290af65de3ea6f15817aa27c5ec7cc5208fd
+checksum="ccbd5a729d730152303be399f033b905e608309d5802d77a61a95faa092592c5
+ f45359be529f5f478ebeb3159859a8f49462df35866e227c12d1440f17451a82
  @a69230c7b6f03a178c93abc6edc832c040d6e198340193b27c3d04afecf8f617"
 skip_extraction="lodepng-${_lodepng_hash}.tar.gz
  skcms-${_skcms_hash}.tar.gz"
diff --git a/srcpkgs/qt-jpegxl-image-plugin/template b/srcpkgs/qt-jpegxl-image-plugin/template
index e9dda7bad0..0193f8b2e6 100644
--- a/srcpkgs/qt-jpegxl-image-plugin/template
+++ b/srcpkgs/qt-jpegxl-image-plugin/template
@@ -1,6 +1,6 @@
 # Template file for 'qt-jpegxl-image-plugin'
 pkgname=qt-jpegxl-image-plugin
-version=0.2.0
+version=0.3.2
 revision=1
 build_style=qmake
 configure_args="qt-jpegxl-image-plugin.pro"
@@ -12,7 +12,7 @@ maintainer="Joshua Krämer <joshua@kraemer.link>"
 license="GPL-3.0-only"
 homepage="https://github.com/novomesk/qt-jpegxl-image-plugin"
 distfiles="https://github.com/novomesk/qt-jpegxl-image-plugin/archive/v${version}.tar.gz"
-checksum=90c8c5b0afb709d4a71f536b3c1eb04d4f26a186220b2cae1b396c42ffbe299c
+checksum=8343500ecb9b92fd1fad5dca74d824ce66991c3eac2ff64e1d5757ef98bfcae8
 
 if [ "$CROSS_BUILD" ]; then
 	hostmakedepends+=" qt5-host-tools"
```

tibequadorian commented:
@mustaqimM It looks like this PR is abandoned and you can take over if you want.

joshuakraemer commented:
Sorry, it's not abandoned, but I'm quite busy, so feel free to take over if you want.

I've updated highway and libjxl to the latest release versions. The patch for libjxl is necessary to make it compile with highway 0.16.0 (changes are cherry-picked from upstream commits). I've dropped qt-jpegxl-image-plugin and instead updated kimageformats to support JPEG-XL.

tibequadorian commented:
Look into plocate template how to fix libatomic linker error.

tibequadorian commented:
If possible, try to wrap lines at 80 characters.

tibequadorian commented on srcpkgs/highway/template:
```diff
@@ -0,0 +1,12 @@
+version=0.16.0
+revision=1
+build_style=cmake
+configure_args="-DBUILD_TESTING=OFF"
```
You may enable/disable tests based on $BUILD_CROSS or $XBPS_CHECK_PKGS.

joshuakraemer commented:
@tibequadorian, thank you for the review and hints. I hope it is ok to merge now.

mustaqimM commented:
@joshuakraemer `kimageformats` now supports jxl. I don't know if you want to swap out `qt-jpegxl-image-plugin` for it:

```diff
diff --git a/srcpkgs/kimageformats/template b/srcpkgs/kimageformats/template
index bc8b9f09a2..58cb9281be 100644
--- a/srcpkgs/kimageformats/template
+++ b/srcpkgs/kimageformats/template
@@ -1,12 +1,12 @@
 # Template file for 'kimageformats'
 pkgname=kimageformats
 version=5.93.0
-revision=1
+revision=2
 build_style=cmake
-configure_args="-DKIMAGEFORMATS_HEIF=ON"
+configure_args="-DKIMAGEFORMATS_HEIF=ON -DKIMAGEFORMATS_JXL=ON"
 hostmakedepends="kcoreaddons extra-cmake-modules qt5-qmake qt5-host-tools
  pkg-config"
-makedepends="karchive-devel libopenexr-devel libheif-devel"
+makedepends="karchive-devel libopenexr-devel libheif-devel libjxl-devel highway"
 short_desc="KDE Plugins to allow QImage to support extra file formats"
 maintainer="John <me@johnnynator.dev>"
 license="LGPL-2.0-or-later"
```

paper42 commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,78 @@
+checksum="ccbd5a729d730152303be399f033b905e608309d5802d77a61a95faa092592c5
+ c47c48c77a205f1af484b7b5a847290af65de3ea6f15817aa27c5ec7cc5208fd
+ @a69230c7b6f03a178c93abc6edc832c040d6e198340193b27c3d04afecf8f617"
+skip_extraction="lodepng-${_lodepng_hash}.tar.gz
+ skcms-${_skcms_hash}.tar.gz"
```
Could we instead use create_wrksrc and build_wrksrc?

joshuakraemer replied:
It would be more complicated because the third-party packages would still need to be moved to specific subdirectories after extraction.

paper42 commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,78 @@
+pkgname=libjxl
+version=0.6.1
+revision=1
+#_short_version=0.6
+#wrksrc=libjxl-${_short_version}
```
these comments should be removed before merging

joshuakraemer replied:
I have removed the two lines. They will likely be needed again though with the next major version (0.7).

tibequadorian replied:
Why? I couldn't find a release tarball where the top directory doesn't match $pkgname-$version

joshuakraemer replied:
Sorry, it seems you're right, I don't know why I had a problem with one of the older versions.

tibequadorian commented:
@joshuakraemer can you resolve conflicts?
@paper42 do you think this is mergeable?

joshuakraemer commented:
> @joshuakraemer can you resolve conflicts?
Done.

joshuakraemer commented:
> @joshuakraemer can you resolve conflicts?

Done.

mustaqimM commented:
Commenting to keep this open.

@joshuakraemer Thank you for keeping up with this, though this should have been merged a while back. There seems to already be `0.7 RC`.

joshuakraemer commented:
I've updated to highway 1.0.1 and libjxl 0.7.0. Unfortunately, the new releases don't build for armv7 (see https://github.com/google/highway/issues/834 and https://github.com/libjxl/libjxl/issues/1748). Even with these issues fixed upstream, it probably won't work for us (without patching at least), because highway seems to expect neon capabilities for the armv7 target, but Void uses `-mfpu=vfpv3`.

classabbyamp commented on srcpkgs/highway/template:
```diff
@@ -0,0 +1,19 @@
+pkgname=highway
+version=1.0.1
+revision=1
+# Build fails for armv7, see https://github.com/google/highway/issues/834
+archs="~armv7l"
```
```suggestion
case "$XBPS_TARGET_MACHINE" in
	armv7*) broken=https://github.com/google/highway/issues/834
esac
```
and move this down below the main block

joshuakraemer replied:
Done.

classabbyamp commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,76 @@
+pkgname=libjxl
+version=0.7.0
+revision=1
+# Build fails for armv7, see https://github.com/libjxl/libjxl/issues/1748
+archs="~armv7l"
```
same thing as comment on highway

joshuakraemer replied:
Done.

classabbyamp commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,76 @@
+	LIBS="-latomic"
+fi
+
+if [ "$CROSS_BUILD" ]; then
```
```suggestion
if [ -z "$XBPS_CHECK_PKGS" ]; then
```

joshuakraemer replied:
Done (and I've also added a condition to only add the pertinent dependencies if tests are built).

classabbyamp commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,76 @@
+checksum="3114bba1fabb36f6f4adc2632717209aa6f84077bc4e93b420e0d63fa0455c5e
+ 1a4d4be37f9a4dd69134832665fda8a772eeb95cbcdaa79d9b1850c23ae50521
+ @a69230c7b6f03a178c93abc6edc832c040d6e198340193b27c3d04afecf8f617"
+skip_extraction="testdata-${_testdata_hash}.tar.gz skcms-${_skcms_hash}.tar.gz"
```
I agree with paper, use `create_wrksrc` and `build_wrksrc` to extract these, then mv them to the right place. it's not more complex, it's more idiomatic within void-packages.

it also means you don't need to depend on tar

joshuakraemer replied:
The skcms archive doesn't have a root directory; therefore, I think it would be necessary to move every single file after automatic extraction with the `create_worksrc` method. Or how would you do it?

classabbyamp commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,76 @@
+
+libjxl-devel_package() {
+	short_desc+=" - development files"
+	depends="libjxl>=${version}_${revision} highway brotli-devel"
```
```suggestion
	depends="${sourcepkg}>=${version}_${revision} highway brotli-devel"
```

joshuakraemer replied:
Done.

classabbyamp commented on srcpkgs/libjxl/template:
```diff
@@ -0,0 +1,76 @@
+	short_desc+=" - tools"
+	pkg_install() {
+		vmove usr/bin
+		vmove usr/share/man
```
```suggestion
		vmove usr/share/man/man1
```
if, e.g., they add man pages in section 3, we don't want them in this package

joshuakraemer replied:
Done.

joshuakraemer commented:
@classabbyamp, thank you very much for your review. Please see my comments above.

classabbyamp commented:
please `git rebase -i --autosquash master` and also update the commit messages and common/shlibs

- updated libjxl and highway (didn't really test this)
- fixed highway on armv7 (libjxl is still broken)
- used checkdepends instead of conditional makedepends
- made it so kimageformats doesn't fail to build on armv7

joshuakraemer commented:
Thank you, @classabbyamp. I hope this will be merged finally.

The failure of the kimageformat test on i686 is not related to this PR, it happens on the master branch as well (@Johnnynator, can you look into it, please?).

classabbyamp commented:
sorry about this taking so long again
- updated highway to fix the armv7 bug
- updated libjxl to fix a CVE
- everything now works on armv7
- fixed extraction for skcms (it needs to be a content checksum `@...` because googlesource generates the tarball in a way that isn't consistent)
- added changelogs

joshuakraemer commented:
@classabbyamp, thank you very much, I'm glad it has been merged finally! :)
